
> Angelina Anaman

> 08/03/2023

> introduction to programming in python

> i6312088


**Introduction**

COVID-19 is a pandemic we never saw coming and a time that has heavily impacted our lives. Understanding how this happened and to what extent different aspects of the pandemic occurred is key to learning from society's mistakes and improving for the future. I notice that the media often glorifies how Scandinavian countries manage society. They often are known for good living and happy society. Therefore I specifically wanted to look at how COVID-19 impacted the different Scandinavian countries and how they dealt with it and if it was effective and efficient. I want to specifically look at the year 2021 because it was the start of the pandemic and will allow me to analyze how they tackled COVID from the start when the disease was not fully understood yet.
This leads me to my research question; How did implementing vaccines and testing influence the covid cases and deaths in Scandinavian countries in 2021?
Questions that will help me develop and answer my research question include:

*   How does vaccination influence deaths?
*   How does testing influence cases?
*   How many tests were taken?
*   What countries are in Scandinavia? 




**Data preparation**

To analyse the data and come to any conclusion I first downloaded the COVID data set from the internet and selected all relevant columns. 

In [1]:
import altair as alt

import pandas as pd

In [2]:
covid = pd.read_csv("https://covid.ourworldindata.org/data/owid-covid-data.csv", parse_dates=['date'])

included_countries = ['NOR', 'DNK', 'SWE', 'GRL', 'FIN', 'ISL', 'FRO','ALA']

included_columns = ['iso_code','date','people_vaccinated','total_deaths','new_deaths','new_cases', 'total_cases','total_tests','new_tests']

covid = covid.loc[covid.iso_code.isin(included_countries) & (covid.date.dt.year == 2021),included_columns]
covid.head()

iso_code       date  people_vaccinated  total_deaths  new_deaths  \
61431      DNK 2021-01-01            31430.0        1322.0        24.0   
61432      DNK 2021-01-02            38469.0        1345.0        23.0   
61433      DNK 2021-01-03            42968.0        1374.0        29.0   
61434      DNK 2021-01-04            45194.0        1389.0        15.0   
61435      DNK 2021-01-05            56370.0        1420.0        31.0   

       new_cases  total_cases  total_tests  new_tests  
61431     2451.0     165930.0   10572230.0    32021.0  
61432     1611.0     167541.0   10645683.0    73453.0  
61433     1170.0     168711.0   10711306.0    65623.0  
61434     2076.0     170787.0   10818667.0   107361.0  
61435     1992.0     172779.0   10924244.0   105577.0

Now that I have my data frame called COVID I can see all the individual countries and the different variables. To confirm all selected coloumns I returned the label of each column in the DataFrame.

In [3]:
columns = list(covid.columns)
columns


['iso_code',
 'date',
 'people_vaccinated',
 'total_deaths',
 'new_deaths',
 'new_cases',
 'total_cases',
 'total_tests',
 'new_tests']

**Explore and Clean Data**

First I wanted to see if there was any missing data in my data frame. I did this by returning the number of missing values in the dataset. As seen below there are no numerical missing values in total total cases and dates. However, a great amount of missing data on people vaccinated. A large number of total tests and total cases are also missing. Inorder to make sure that no wrong conclusions drawn because of the missing date. I explicitly indicated the data type of interest in the data frame. Additionally, having missing data will make the graphs inconsistent, which will hinder my understanding of the different relationships. Since the iso code is not a numerical value check if there are any missing values. Therefore I have to specify that it is an object.


In [4]:
covid.isnull().sum()

iso_code                0
date                    0
people_vaccinated    1358
total_deaths          365
new_deaths            370
new_cases               3
total_cases             0
total_tests           724
new_tests             731
dtype: int64

In [5]:
covid.describe(include=['object'])

iso_code
count      2555
unique        7
top         DNK
freq        365

In my opinion in this case outliers can be very signifanct and show me how some variables reached an extreme. To find the outliers I use the box plot method to see any extreme values.

In [6]:
bptest = alt.Chart(covid).mark_boxplot(extent='min-max').encode(
    x='iso_code:O',
    y='new_tests:Q'
)
bptest

alt.Chart(...)

In [7]:
bpcase = alt.Chart(covid).mark_boxplot(extent='min-max').encode(
    x='iso_code:O',
    y='new_cases:Q'
)

bpcase

alt.Chart(...)

In [8]:
bpdeath = alt.Chart(covid).mark_boxplot(extent='min-max').encode(
    x='iso_code:O',
    y='new_deaths:Q'
)

bpdeath

alt.Chart(...)

In [9]:
bpvac = alt.Chart(covid).mark_boxplot(extent='min-max').encode(
    x='iso_code:O',
    y='people_vaccinated:Q'
)

bpvac

alt.Chart(...)

The box plots above help me understand the outliers. For example, in Sweden, the record number of new deaths a day was almost 500. This is a large difference compared to other Scandinavian countries. Creating box plots also allowed me to understand that Denmark was very efficient in testing as the most tests reported in a day were around 170,000. A describe function was created before cleaning the data to compare the number of values left after cleaning. 

In [10]:
covid.describe()

people_vaccinated  total_deaths   new_deaths     new_cases  \
count       1.197000e+03   2190.000000  2185.000000   2552.000000   
mean        1.874069e+06   3058.324658     4.810984    833.255486   
std         1.948722e+06   4994.408635    22.023224   2306.456615   
min         6.700000e+01      1.000000     0.000000      0.000000   
25%         4.542200e+04     30.000000     0.000000      0.000000   
50%         9.329290e+05    871.000000     0.000000    138.500000   
75%         4.086473e+06   2535.750000     2.000000    679.000000   
max         7.636790e+06  15310.000000   474.000000  41035.000000   

        total_cases   total_tests      new_tests  
count  2.555000e+03  1.831000e+03    1824.000000  
mean   2.250847e+05  1.089044e+07   36934.945175  
std    3.441818e+05  1.253971e+07   49299.221553  
min    2.700000e+01  2.414960e+05      42.000000  
25%    1.688500e+03  3.337178e+06    9207.500000  
50%    8.424700e+04  6.725499e+06   18763.000000  
75%    2.586190e+05  1.197605e+07   37698.000000  
max    1.314784e+06  5.356980e+07  269142.000000

In [11]:
covid.dropna(inplace=True)
print(covid.isnull().sum())

iso_code             0
date                 0
people_vaccinated    0
total_deaths         0
new_deaths           0
new_cases            0
total_cases          0
total_tests          0
new_tests            0
dtype: int64


In [12]:
covid.describe()

people_vaccinated  total_deaths  new_deaths     new_cases  \
count       9.430000e+02    943.000000  943.000000    943.000000   
mean        2.359898e+06   2082.785790    3.042418   1043.520679   
std         1.913810e+06   3031.801911    7.101352   2342.463425   
min         6.700000e+01     29.000000    0.000000      0.000000   
25%         3.014865e+05    734.500000    0.000000    202.500000   
50%         2.165702e+06    999.000000    0.000000    484.000000   
75%         4.219310e+06   2533.500000    3.000000    907.500000   
max         7.636790e+06  15265.000000   61.000000  41035.000000   

        total_cases   total_tests      new_tests  
count  9.430000e+02  9.430000e+02     943.000000  
mean   2.392778e+05  1.572775e+07   54359.989396  
std    2.320784e+05  1.540517e+07   61659.761496  
min    5.975000e+03  2.552910e+05     363.000000  
25%    9.673800e+04  4.705840e+06   10357.000000  
50%    1.939170e+05  7.952429e+06   23379.000000  
75%    3.081895e+05  2.870078e+07   80264.000000  
max    1.273313e+06  5.356980e+07  269142.000000

After cleaning the data frame and deleting all the missing values I was left with less data points seen in the count of the new describe function. To check that there a no longer missing values in my data frame I checked it the isnull method.

**Describe and Visualize**

To explore my research question more I decided to explore the population of my chosen data set more. As mentioned before I chose Scandinavian countries specifically, but who are these people really of the data set? Describing the population with other factors allows me to understand what other factors influenced the way governments handled the pandemic.


In [13]:
descriptive_stat = pd.read_csv("https://covid.ourworldindata.org/data/owid-covid-data.csv", parse_dates=['date'])

included_countries = ['NOR', 'DNK', 'SWE', 'GRL', 'FIN', 'ISL', 'FRO','ALA']

included_columns = ['iso_code','continent','gdp_per_capita','extreme_poverty','female_smokers','male_smokers','population']

descriptive_stat = descriptive_stat.loc[descriptive_stat.iso_code.isin(included_countries) & (descriptive_stat.date.dt.year == 2021),included_columns]
descriptive_stat.head()

iso_code continent  gdp_per_capita  extreme_poverty  female_smokers  \
61431      DNK    Europe       46682.515              0.2            19.3   
61432      DNK    Europe       46682.515              0.2            19.3   
61433      DNK    Europe       46682.515              0.2            19.3   
61434      DNK    Europe       46682.515              0.2            19.3   
61435      DNK    Europe       46682.515              0.2            19.3   

       male_smokers  population  
61431          18.8   5882259.0  
61432          18.8   5882259.0  
61433          18.8   5882259.0  
61434          18.8   5882259.0  
61435          18.8   5882259.0

In [14]:
descriptive_stat.describe()

gdp_per_capita  extreme_poverty  female_smokers  male_smokers  \
count     1825.000000      1460.000000     1825.000000   1825.000000   
mean     49100.106800         0.275000       18.060000     19.240000   
std       8203.509489         0.129948        1.931954      2.452284   
min      40585.721000         0.200000       14.300000     15.200000   
25%      46482.958000         0.200000       18.300000     18.800000   
50%      46682.515000         0.200000       18.800000     18.900000   
75%      46949.283000         0.275000       19.300000     20.700000   
max      64800.057000         0.500000       19.600000     22.600000   

         population  
count  2.555000e+03  
mean   3.984170e+06  
std    3.687595e+06  
min    5.311700e+04  
25%    5.649400e+04  
50%    5.434324e+06  
75%    5.882259e+06  
max    1.054935e+07

In [15]:
alt.Chart(descriptive_stat).mark_bar().encode(
    x='iso_code',
    y='mean(population)'
)

alt.Chart(...)

With a new data frame to analyse the population I realized that scandinvaian countires  have very low extreme poverty levels. Additionally the mean smokers are not very much this means that societies have very healthy lungs.
Below I analysed addtional statistics of the dataframe to understand the most cases scandnavain coutnires had in one day. It is also seen that in Scandinavian countries they have a high gdp per captia of aorund 18,00 when the averge in the work is 12,000. This can perhaps show that more money influenced the resources made available to handle the pandemic is 2021.  Overall, Sweden has a higher population than other countries which will help my analysis further on. 

Below I tested random descriptive stastics like the max and the mean to better understand my data set. For example, I see that the max covid cases in all countires was 41035. 

In [16]:
covid['new_cases'].max()

41035.0

In [17]:
covid['new_cases'].mean()

1043.5206786850476

In [18]:
covid['people_vaccinated'].max()

7636790.0

In [19]:
covid['new_tests'].max()

269142.0

In [20]:
covid['new_tests'].mean()

54359.98939554613

To make the relationships more visual and how the different variables changed in 2021 line graphs were made. First individual graphs of each chosen variable. 

In [21]:
date_case = alt.Chart(covid).mark_line().encode(
    x='date',
    y='total_cases',
    color='iso_code'
)
date_case

alt.Chart(...)

In [22]:
date_test = alt.Chart(covid).mark_line().encode(
    x='date',
    y='total_tests',
    color='iso_code'
)
date_test

alt.Chart(...)

In [23]:
date_death = alt.Chart(covid).mark_line().encode(
    x='date',
    y='total_deaths',
    color='iso_code'
)
date_death

alt.Chart(...)

In [24]:
date_vaccine = alt.Chart(covid).mark_line().encode(
    x='date',
    y='people_vaccinated',
    color='iso_code'
)

date_vaccine

alt.Chart(...)

To test the two relationships that will help me answer my researh question I decided to see if testing influenced the amount of cases and if the amount of people vaccinated affected the amount of deaths. With this an interactive widget was used to see how the relationships varied in the Scandinavian countries.

In [25]:
import ipywidgets as widgets


In [26]:
widgets.Dropdown(options=['NOR', 'DNK', 'SWE', 'GRL', 'FIN', 'ISL', 'FRO','ALA'])

Dropdown(options=('NOR', 'DNK', 'SWE', 'GRL', 'FIN', 'ISL', 'FRO', 'ALA'), value='NOR')

In [27]:
def select_country_data(iso_code, df):  
  return df.loc[df['iso_code'] == iso_code, ]

In [28]:
import altair as alt

def visualize_country_data(iso_code, df):
  data_to_viz = select_country_data(iso_code, df)
  chart = alt.Chart(data_to_viz).mark_line().encode(
      x='total_tests',
      y='total_cases',
      tooltip='iso_code'
  )

  display(chart)

In [29]:
from ipywidgets import interact

In [30]:
from ipywidgets import fixed

In [31]:
def f(x):
  return

In [32]:
interact(f, x=['NLD', 'GBR', 'BEL']);

interactive(children=(Dropdown(description='x', options=('NLD', 'GBR', 'BEL'), value='NLD'), Output()), _dom_c…

In [33]:
interact(f,
         x=[('Sweden', 'SWE'),
            ('Denmark', 'DNK'),
            ('Norway', 'NOR'),
            ('Finland', 'FIN'),
            ('Iceland','ISL')]);

interactive(children=(Dropdown(description='x', options=(('Sweden', 'SWE'), ('Denmark', 'DNK'), ('Norway', 'NO…

In [34]:
interact(visualize_country_data,
         iso_code=[('SWE'),
                                        ('NOR'),
                                        ('DNK'),
                                        ('FIN')],
         df=fixed(covid));

interactive(children=(Dropdown(description='iso_code', options=('SWE', 'NOR', 'DNK', 'FIN'), value='SWE'), Out…

In [35]:
def select_country_data(iso_code, df):  
  return df.loc[df['iso_code'] == iso_code, ]

In [36]:
import altair as alt

def visualize_country_data(iso_code, df):
  data_to_viz = select_country_data(iso_code, df)
  chart = alt.Chart(data_to_viz).mark_line().encode(
      x='people_vaccinated',
      y='total_deaths',
      tooltip='iso_code'
  )

  display(chart)

In [37]:
interact(visualize_country_data,
         iso_code=[('SWE'),
                                        ('NOR'),
                                        ('DNK'),
                                        ('FIN')],
         df=fixed(covid));

interactive(children=(Dropdown(description='iso_code', options=('SWE', 'NOR', 'DNK', 'FIN'), value='SWE'), Out…

**Discussion and Conclusion**

Through multiple steps, I used a large data set to analyze how various Scandinavian countries dealt with the infamous pandemic. Looking at the data represented visually total cases were most in Sweden and all countries increased in 2021. Finland, Iceland, and Denmark had a very slight increase in total cases. Even though Sweden had the most cases Denmark has more total tests over the year 2021. On the other hand, Iceland used fewer tests throughout 2021. I think this could be due to the population. Sweden has the highest population and therfore will significantly be higher. Total deaths plateaued for many countries throughout the year. Sweden had the most total deaths which are alarming compared to the other countries. According to the graph, almost all Scandinavian countries implemented vaccination in March. There are large increases of people having vaccinations and ended in a plateau at the end of the year. This could indicate that the majority of the population has already been vaccinated. 

To see how different governmental methods (testing and vaccinations) influenced deaths and cases interactive visualizations were used. Overall, in all countries, more testing resulted in more cases. Additionally, there is a significant pattern in the relationship between vaccination and death in all countries. There is a large plateau in the middle of 2021. This showed that when people were being vaccinated there was almost no increase in death. This shows these countries valued implementing vaccines to prevent death. Therefore looking at is statistically, the relationship between cases and test and vaccines and deaths were all postivie. However the trend in which they increased showed that the countries used alot of vaccination. From analysis of the population, this could be due to a high gdp per capita. Addiotionly, more testing resulted to more cases and through graphs total tests increased throughoout 2021. This shows that they also took testing seriously.  